Занесение всех моделей проекта в дерево сериализованных объектов

bogachev-pa opened the pull request.

bogachev-pa committed:
adding all rao editors in project to serialization config view

related issue: aurusov/rdo-xtext#107

bogachev-pa committed:
serialization config: provide minimal api (aurusov/rdo-xtext#107)

bogachev-pa committed:
serialization config: watch for file changes in inactive editors

related issue: aurusov/rdo-xtext#107

aurusov commented on ru.bmstu.rk9.rao.ui/src/ru/bmstu/rk9/rao/ui/serialization/SerializationConfigView.java:
```diff
@@ -22,37 +37,90 @@
+	 *
+	 * @return true on success, false if failed to uncheck all elements (tree is
+	 *         not yet initialized)
+	 */
```
Зачем документация ?

bogachev-pa replied:
Потому что это вроде как `API`.

aurusov replied:
Не поддерживаю введение java-doc даже ради API. Для понятного API код должен быть читым, этого вполне достаточно.

bogachev-pa replied:
Мне лично всегда удобнее посмотреть описание метода либы в джавадоке, чем лезть в его исходники и разбираться там. Благо эклипс очень хорошо с ними работает.

Поясните, почему вы против?

aurusov replied:
Потому что я так делал в старом РДО. Потом находил расхождение комментария с кодом (код развивается !!). Потом потратил время на удаление комментариев. Не хочу начинать заново.

bogachev-pa replied:
Это неправильный подход. Правильный - следить за тем, чтобы комментарии менялись при изменении кода.

aurusov replied:
Т.е. твой правильный подход - тратить ресурсы. _Следить_ и _править_ стоят ресурсов. У меня их нет. Что делать ?

bogachev-pa replied:
Ну, мои аргументы такие:
1. Пользоваться документированными методами намного приятнее.
2. Следить за внесением исправлений в документацию гораздо проще теперь, когда у нас есть гитхаб и оформленный гитфлоу, чем раньше под `svn`.

Если вас эти аргументы ни в чем не убеждают, то ок, я спилю.

aurusov replied:
1. Если код чистый, то это спорный вопрос. Вот пример, когда без комментария смотрится лучше https://github.com/aurusov/rdo_studio/commit/4736af2c79e05776502ceac115025929ecdfc99d#diff-ce19d6ed7dab275c18b1b908b633c36aL80. Понятно, что это imho. А как тебе это https://github.com/aurusov/rdo_studio/commit/4736af2c79e05776502ceac115025929ecdfc99d#diff-89876a411684756c26ce0b7070ba2650L123 ?
2. Проверь https://github.com/aurusov/rdo-xtext/pull/110#commitcomment-12755984

aurusov commented on ru.bmstu.rk9.rao.ui/src/ru/bmstu/rk9/rao/ui/serialization/SerializationConfigView.java:
```diff
@@ -22,37 +37,90 @@
 	// ――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――― //
-	// ---------------------------- VIEW SETUP ----------------------------- //
+	// ------------------------------ VIEW SETUP --------------------------- //
 	// ――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――― //
```
Ну началось. Не надо так больше делать.

bogachev-pa replied:
Случайно вышло, ок.

aurusov commented on ru.bmstu.rk9.rao.ui/src/ru/bmstu/rk9/rao/ui/serialization/SerializationConfigView.java:
```diff
@@ -22,37 +37,90 @@
+	 */
+	public final boolean checkAll() {
+		return setStateForAll(true);
+	}
```
Где используется возвращаемое значение ? Если не используется, то `void`.

bogachev-pa replied:
Эти методы вообще не используются. В сообщению к коммиту же написано - это `API`. Пригодится плагину пятнашек.

aurusov replied:
Тогда `void`. Если плагину потребуется, то тот, кто это будет делать, изменит и тут.

bogachev-pa replied:
Если метод может пофейлиться, то это как-то неправильно, что он возвращает `void`. Он должен либо уметь сообщать об ошибке, либо выкидывать исключение (что не кажется мне очень хорошей практикой).
Почему для `API` мы должны делать исключение и заведомо закладывать плохой дизайн?

aurusov replied:
Как раз сейчас дизайн _избыточен_, значит плохой. Как только потребуется, мы внесем изменение в API. Мы можем себе такое позволить, потому что этим API никто не пользуется. Когда оно устаканется на практике, тогда и будем его фризить и ломать голову над обратной совместимостью.

bogachev-pa replied:
Ок.

aurusov commented on ru.bmstu.rk9.rao.ui/src/ru/bmstu/rk9/rao/ui/serialization/SerializationConfigView.java:
```diff
@@ -100,163 +168,307 @@ public void doubleClick(DoubleClickEvent event) {
 						}
 					});
+				} catch (CoreException e) {
+					e.printStackTrace();
```
А почему дальше не пробрасывается ? Так и задумано ?

bogachev-pa replied:
Пока не очень представляю, как исключения тут можно правильно похэндлить.

aurusov replied:
Опиши словами проблему.

bogachev-pa replied:
Интерфейс описан как `throws`, но если полезть в исходники, то, во-первых, исключение никак не может быть вызвано неправильными данными от пользователя, т.е. это всегда какая-то ошибка при работе с файлами самого эклипса, а, во-вторых, в имплементации оно там везде хэндлится внутри (если я ничего не упустил, конечно), и наружу никак не попадет.

 В итоге непонятно, что все-таки случилось, если исключение тут вдруг было выброшено, и как это похэндлить кроме как проигнорировать.

aurusov replied:
Тогда оформи это логирование как warning.

bogachev-pa replied:
В смысле? Сейчас в лог ошибки выведется в консоль. Еще что-то нужно?

aurusov replied:
> В смысле? Сейчас в лог ошибки выведется в консоль. Еще что-то нужно?

И это невозможно использовать от слова совсем. Мне приходится плясать с бубном, чтобы понять какой стек к чему относится. Так что текущий уровень логирования неудовлетворительный. Помимо стека нужен контекст,
- рукотворное сообщение об ошибке
- категория к каждому логу (info/warning/error)
- уровень лога (0-10)

Логи должны иметь стандартный формат, чтобы их грепать. Надо заюзать какую-то готовую систему логирования.

http://ru.stackoverflow.com/questions/332814/%D0%9B%D0%BE%D0%B3%D0%B8%D1%80%D0%BE%D0%B2%D0%B0%D0%BD%D0%B8%D0%B5-%D0%B2-java

bogachev-pa replied:
Завел под это отдельную таску (https://github.com/aurusov/rdo-xtext/issues/120) , предлагаю перенести обсуждения туда и оформлять это отдельным пул-реквестом.

aurusov replied:
Хорошо. Остались только комментарии ?

bogachev-pa replied:
Еще уберу исключения в пользу методов с двумя параметрами.

bogachev-pa committed:
serialization config: proper tree updating (aurusov/rdo-xtext#107)

- forming tree from all opened projects
- fix updating during files creation/deletion

bogachev-pa committed:
serialization config: simplify api (aurusov/rdo-xtext#107)

aurusov commented on ru.bmstu.rk9.rao.ui/src/ru/bmstu/rk9/rao/ui/serialization/SerializationConfigView.java:
```diff
@@ -22,37 +35,83 @@
+	 */
+	public final void uncheckAll() {
+		setStateForAll(false);
+	}
```
Возможно, ты держишся за комментарии, потому что тебе не нравятся названия методов ? Может быть
`setCheckedStateForAll()` и `unsetCheckedStateForAll()`

bogachev-pa replied:
Ок, сделал так.

bogachev-pa committed:
serialization config: slight code refactoring (aurusov/rdo-xtext#107)

bogachev-pa committed:
serialization config: rename api methods and remove java docs

related issue: aurusov/rdo-xtext#107
